Objetivo 2 [IV-22-23]

JoseCarlosPPK opened the pull request:
## Sobre la estructura del repositorio
* [x] ¿He seguido las mejores prácticas en nombre de las clases y ficheros y
   disposición de los mismos?
- El nombre de los ficheros están en minúscula para evitar problemas con otros SOs y *filesystems* que no distinguen entre mayúculas y minúsculas. En JS se usa camelCase (variables, funciones), PascalCase para clases y todo en mayúsculas para constantes

## Sobre el análisis del problema
* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo
   que se ha creado?
- En el cuerpo de los issues se habla del análisis realizado.
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor,
   una entidad o un agregado?
Lo respondo aquí mismo. El modelo se basa en:
1. Los objeto-valor Libro y ConjuntoGeneros. Estos son inmutables. Cuando se crea un objeto libro sus propiedades no pueden modificarse y lo mismo pasa cuando creas un objeto de ConjuntoGeneros.
2. La entidad Usuario. Es una entidad porque sus gustos pueden variar.

* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo
   objeto valor que no dependa de ningún otro (y que sea la base de muchos
   otros)?.
Creo que sí he propuesto un PMV. Los gustos del usuario se basan en los géneros de los libros y de esta manera se pueden realizar comparaciones para poder luego recomendar libros. 
El objeto-valor ConjuntoGeneros no depende de ningún otro y es la base del objeto-valor Libro y la entidad Usuario

## Sobre la planificación y la programación
* [x] ¿Los issues representan un problema, y no una tarea?
* [x] ¿Todos los mensajes de commit explican el cambio, y no se
      limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
- Mayoritariamente sí.
La primera línea es <= a 50 caracteres. Se indica un verbo en imperativo que indica la acción realizado.
Se ha escrito un cuerpo del commit que describa con más detalle el cambio realizado.
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos
   los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están asignados a una HU?
* [ ] ¿Todos los cambios en el código están asignados a un issue al que se
      referencia en un commit?
- Hay algunos que no referencian al issue. Esto es debido a que era un cambio de nombre, añadir comentarios al  código... y no sabía cómo asociarlo o no me parecía correcto. He asociado un commit a un issue cuando se añadía/modificaba código.
* [x] Los issues sobre los que estoy trabajando, ¿han
      sido asignados al primer milestone?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Son los milestones sobre los que estoy trabajando productos
   mínimamente viables? ¿O tengo que solicitar al product manager que
   cree milestones adicionales o precise de qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente
correcto?
Con node --check <archivo.js> se puede comprobar la sintaxis y no hay ningún problema.


Close #9 porque está mal planteado y no se va a seguir con su desarrollo.
Close #11 porque no se va a seguir con su desarrollo.

JoseCarlosPPK committed:
Elegido lenguaje de programación close #8

JoseCarlosPPK committed:
Clasificación propia de los libros, close #10

Se han añadido varias clasificaciones para los libros. Se ha usado un objecto de Javascript con pares clave-valor, emulando un enumerado. De esta manera podemos usar la notación punto (.) para acceder a las claves (comodidad). Se usa el objeto Symbol para obtener valores únicos, para evitar un posible error futuro si se hacen comparaciones entre distintos enumerados que por casualidad tienen valores iguales

JoseCarlosPPK committed:
docs(M0): añade documentación para la propuesta de #10

Se analiza el problema general y documenta la solución propuesta en src/model/generos.js

JoseCarlosPPK committed:
refactor: añade comentarios al código

JoseCarlosPPK committed:
Add representación de un libro y close #12

Los libros para trabajar con la lógica de negocio recogen los géneros a los que pertenece y un identidicador para distinguir el libro en cuestión

JoseCarlosPPK committed:
Add asociación libro-puntuación; close #13

Se relaciona un id de un libro con un valor numérico que representa la puntuación

JoseCarlosPPK committed:
Change nombre de fichero

Nombre de fichero más representativo

JoseCarlosPPK committed:
Delete una línea de código sobrante

En una primera versión se usaba la clase Libro y luego se decidió usar el id, por lo que no hace falta importar la clase Libro

JoseCarlosPPK committed:
Add entrada entidad en iv.yaml

Considero que PuntuacionLibro es una entidad ya que depende
de la puntuación que la lógica de negocio calcule

JoseCarlosPPK committed:
Add entrada entidad vacía

Considero que lo creado en el modelo son objetos-valor y no hay ninguna entidad

JoseCarlosPPK committed:
Merge branch 'objetivo-2' of github.com:JoseCarlosJC/Akinah into objetivo-2

JoseCarlosPPK commented:
En estos dos últimos commits la he liado un poco. El commit [d6a3b8b](https://github.com/fjgallardo00/Akinah/pull/14/commits/d6a3b8bf18bfa523db23e75816f860f6b1fb6339) lo quería subir sin una entrada para entidad en el fichero iv.yaml y subí un fichero desactualizado

JJ commented on src/model/generos.js:
```diff
@@ -0,0 +1,15 @@
+
+
+/** Gustos del usuario: será un subconjunto de generos */
+export const gustos = new Set()
```
Esto es una variable global, y además constante. Es imposible que lo asocies a un solo usuario.

JJ commented on src/model/libro.js:
```diff
@@ -0,0 +1,28 @@
+ * Asocia un identificador de un libro con los géneros que lo clasifican
+ */
+export class Libro{
+   #id
```
Habrá que poner alguna información además del ID, ¿no?

JJ reviewed (changes requested):
Tendrás que reflejar un poco mejor las historias de usuario, sobre todo en lo que concierne al mismo usuario.

pedrooot commented:
Como ha comentado @JJ, los libros tienen más atributos de los que estás creando en libro.js (título...). 
Tampoco me queda muy claro el tema de la puntuación ya que esta se asigna dependiendo de los gustos de cada usuario y la implementación actual de la misma toma la puntuación como algo general sin tener en cuenta el usuario (Para cada libro, la puntuación del mismo es distinta dependiendo de los gustos del usuario).  
Intenta ir revisando estas cosas y vamos viendo 😬

JoseCarlosPPK committed:
Delete uso de 'use strict' en JS

Desde ES2015 no es necesario ponerlo explícitamente

JoseCarlosPPK committed:
Add modelo de los géneros y close #10

En lugar de tener un conjunto que represente la clasificación propia de los géneros, se ha optado por crear una clase que contenga de manera general todos los géneros y las instancias de dicha clase sean subconjuntos de este primero

JoseCarlosPPK committed:
Add modelo de los usuarios; close #15

Se crea la clase Usuario que encapsula los gustos de este. Los gustos vienen representados por un objeto ConjuntoGeneros

JoseCarlosPPK committed:
Add modelo de los libros; close #12

Se define una clase que recoge el título, autor, idioma, descripción y géneros de un libro.
Se ha decidido incluir la descripción ya que si se recomiendan varios libros, el usuario podría hacerse una idea
de qué va cada libro y así poder escoger

JoseCarlosPPK committed:
Add modelo del libro sin idioma; close #12

Se decide borrar el idioma de los libros. No se tiene en cuenta en los usuaros así que no se debería de tener en cuenta aquí. Como producto mínimamente viable nos estamos centrando solo en los géneros. El idioma podría afectar a las recomendaciones que realice la lógica de negocio. A diferencia de la descripción, esta la tendrá en cuenta solo el usuario si se realiza más de una recomendación

JoseCarlosPPK committed:
Delete modelo de la puntuación; close #13

Crear una clase que relacione la puntuación con un libro tenía el objetivo de poder ordenar aquellos libros con mayor puntuación. Esto sería parte de la lógica de negocio

JoseCarlosPPK committed:
Add entidades al fichero iv.yaml

Las clases Usuario y ConjuntoGeneros han de ser entidades ya que sus valores pueden cambiar: un usuario puede cambiar de gustos y un conjunto se puede modificar

JoseCarlosPPK commented:
Listo para revisión @fjgallardo00 !

JoseCarlosPPK commented:
Perdón por cerrar el PR, ha sido sin querer. Me he confundido con *cancelar* el comentario que iba a escribir

JoseCarlosPPK committed:
Refactor nombre de un archivo

generos.js pasa a ser conjunto_generos.js, como el nombre de la clase

JoseCarlosPPK commented:
El resumen de los cambios:
1. Los géneros se han modelado como una clase que como atributo de clase tiene el conjunto de géneros global y cada instancia será un subconjunto de este.
2. Se han modelado los usuarios. Cada usuario encapsula sus gustos, que son objetos de la clase ConjuntoGeneros.
3. Se ha eliminado la clase que relacionaba la puntuación con los libros. Como bien dijo @pedromarting3 , el modelo anterior representaba una puntuación global para todos los usuarios para un mismo libro. Una vez modelados los usuarios, para solucionar el problema comentado por @pedromarting3, en vez de tener la pareja (libro, puntuación) se podría tener la terna (usuario, libro, puntuación), pero la idea de esta clase era ayudar a la lógica de negocio cuando tuviera que ordenar los libros con mayores puntuaciones. Es aquí cuando no me queda claro si esta clase debería formar parte del modelo y por eso la he quitado, pensando que es algo más interno de la lógica de negocio

JoseCarlosPPK committed:
Change ConjuntoGeneros a objeto-valor

He decidido que cada conjunto de géneros sea inmutable. Si quieres modificar uno tendrás que crear un nuevo objeto

JoseCarlosPPK commented:
Con esto estaría listo para revisión @fjgallardo00

fjgallardo00 commented:
@JoseCarlosJC el archivo de iv.yaml le falta un retorno de carro. No te pasarán los tests de JJ si no se lo pones. Por lo demás lo veo todo bien

JoseCarlosPPK committed:
Add nueva línea en blanco al final de iv.yaml

JoseCarlosPPK commented:
Le he añadido esa nueva línea al final aunque no se muestra aquí una tercera línea

JoseCarlosPPK commented:
Listo para revisión @JJ . No puedo volver a pasar los tests porque me aprobó el PR y no debería abrir otro para el mismo objetivo según *las normas*

JJ commented on src/model/conjunto_generos.js:
```diff
@@ -0,0 +1,69 @@
+    * los valores que se encuentren en @see GENEROS
+    * @post Los géneros serán un subconjunto de @see GENEROS
+    */
+   #setGeneros(conjunto){
```
Si esto lo llamas solo desde el constructor, ¿para qué lo quieres crear? Además, así dejas claro que sólo el constructor puede establecer el valor de esa variable y lo haces inmutable. Si no lo haces así, no sería un objeto valor, puesto que cualquier podría llamar a esta función privada.

JJ commented on src/model/conjunto_generos.js:
```diff
@@ -0,0 +1,69 @@
+    * @post Los géneros serán un subconjunto de @see GENEROS
+    */
+   #setGeneros(conjunto){
+      this.#generos.clear()
```
Deberíais acostumbraros a crear el código de acuerdo a los requisitos y la lógica de negocio. Una vez más, estás haciendo el objeto mutable mediante un setter. Si lo que quieres es mutar el objeto, primero, no sería un objeto valor; segundo, debes ver la mejor forma de hacerlo (con operaciones CRUD, por ejemplo)

JJ commented on src/model/libro.js:
```diff
@@ -0,0 +1,48 @@
+export class Libro{
+   #titulo
+   #autor
+   #descripcion
```
Repetidamente os he dicho que debéis buscar el producto *mínimamente* viable, y por supuesto que se ajuste a la historia de usuario. En ningún lugar en #7 ni en #12 (que, por cierto, no plantea un problema) dice lo que comentas en db371ce9ac6338aa07adad9de12c0b84e8875c24, "para que se haga una idea". La definición de deuda técnica es incluir código que no se ha pedido en ningún lugar como "solución" a un problema que en realidad no existe. Por favor, no lo hagáis.

JoseCarlosPPK replied:
> Repetidamente os he dicho que debéis buscar el producto _mínimamente_ viable, y por supuesto que se ajuste a la historia de usuario. En ningún lugar en #7 ni en #12 (que, por cierto, no plantea un problema) dice lo que comentas en [db371ce](https://github.com/fjgallardo00/Akinah/commit/db371ce9ac6338aa07adad9de12c0b84e8875c24), "para que se haga una idea". La definición de deuda técnica es incluir código que no se ha pedido en ningún lugar como "solución" a un problema que en realidad no existe. Por favor, no lo hagáis.

Aquí es donde dice @JJ que #12 no plantea un problema (#7 es la HU). Por eso me replanteé si el error estaba en la manera de expresarlo. Por el resto tiene razón y ya he aprendido que si hay un issue incorrecto o innecesario, se borra el código añadido por ese issue y se cierra en el commit

JJ reviewed (changes requested):
Está bastante mejor y has seguido de forma más o menos adecuada las buenas prácticas. Desgraciadamente, el problema fundamental, que es desarrollar sólo el código que te pidan en los issues, no lo has hecho.
Adicionalmente, mira los problemas que se te han comentado.

JoseCarlosPPK committed:
Delete método setter de ConjuntoGeneros

Solo será posible asignar un valor al atributo desde el constructor

JoseCarlosPPK committed:
Change atributo de clase GENEROS; close #10

Aunque fuera un dato privado, el Set GENEROS podía modificarse dentro de los métodos y no es lo que queremos. Para hacer un set inmutable no sirve Object.freeze(). Puede que haya maneras más complejas de conseguir que un set sea inmutable, o usar código extero como *inmutable.js*. La solución propuesta es definir un método estático que devuelve una propiedad (get). Cada vez que se llama se crea el conjunto de GENEROS. Esta es la única manera de acceder al conjunto y por tanto si vuelves a consultar su valor, al crearse uno nuevo y siempre el mismo, se obtiene el efecto de que no se puede modificar

JoseCarlosPPK committed:
Change Usuario para que trabaje con ConjuntoGeneros; close #15

Al realizar el cambio de nombre del fichero de la clase ConjuntoGeneros, no cambié el import en este

JoseCarlosPPK committed:
Delete atributo descripcion; close #12

En ningún lado de la HU1 #7 se meciona la descripción, solo los gustos del usuario, que por la definición del problema, se basan en los géneros

JoseCarlosPPK committed:
Fix import de libro.js; close #12

Al cambiar el nombre del fichero de los géneros no se realizó el cambio del import en este fichero

JoseCarlosPPK commented:
Listo para revisión @fjgallardo00  @JJ! 
He quitado el setter y el código de asginación al atributo solo lo ejecuta el constructor. Además como mejora (pienso), ningún objeto de la clase puede modificar el atributo de clase GENEROS. 
Para los libros he quitado el atributo *descripcion* que no se menciona en ningún issue y no es necesario para el PMV.

El issue #12 he reformulado el título para que exprese un problema, no sé si el error estaba ahí, en la expresión. El problema es que se han de modelar los libros porque la lógica de negocio los usa para sus cálculos y finalmente como recomendaciones (esto está expresado en el issue)

JJ commented:
> Listo para revisión @fjgallardo00 @JJ! He quitado el setter y el código de asginación al atributo solo lo ejecuta el constructor. Además como mejora (pienso), ningún objeto de la clase puede modificar el atributo de clase GENEROS. Para los libros he quitado el atributo _descripcion_ que no se menciona en ningún issue y no es necesario para el PMV.
> 
> El issue #12 he reformulado el título para que exprese un problema, no sé si el error estaba ahí, en la expresión. El problema es que se han de modelar los libros porque la lógica de negocio los usa para sus cálculos y finalmente como recomendaciones (esto está expresado en el issue)

Repetidamente os he dicho que reformular un issue es una mala práctica. La buena práctica es cerrarlo cuando se borra el código que se ha creado con él y crear un nuevo issue.

JJ commented on src/model/usuario.js:
```diff
@@ -0,0 +1,37 @@
+    * Cambia los gustos del usuario
+    * @param {ConjuntoGeneros} nuevosGustos
+    */
+   set gustos(nuevosGustos){
```
Primero, en este milestone se trata solo de modelizar, no de añadir funcionalidad. Por otro lado, repetidamente os he dicho en clase que *no* pongáis setters. El API de una clase debe regular cómo muta, y si lo hace o no. Si es mutable, ya no es un objeto valor, por otro lado, pero en todo caso, tendrá que mutar solamente de las formas que permita el interfaz.
Todo esto es consecuencia de que has hecho el cambio en 670e03ced7b6c6c184b9b5b5f3b7d82a0cde7ee0 que no apunta a ningún issue, *y por tanto es incorrecto*.

JJ commented on src/model/libro.js:
```diff
@@ -0,0 +1,42 @@
+   constructor(titulo, autor, generos){
+      this.#titulo = titulo
+      this.#autor = autor
+      this.#generos = generos
```
¿Dónde estás usando `ConjuntoGeneros`?

JJ commented on src/model/usuario.js:
```diff
@@ -0,0 +1,37 @@
+    * @param {ConjuntoGeneros} gustos - Conjunto de gustos
+    */
+   constructor(gustos){
+      this.#gustos = gustos
```
¿Donde compruebas que esto es del tipo correcto? Dicho de otro modo, ¿por qué importas la clase correspondiente si no lo usas?

JoseCarlosPPK replied:
Entiendo

JJ commented on src/model/conjunto_generos.js:
```diff
@@ -0,0 +1,53 @@
+/**
+ * @class ConjuntoGeneros
+ */
+export class ConjuntoGeneros{
```
Esto es, al final del día, un `Set` que no tiene existencia indepenciente, ni para el caso funcionalidad, fuera de los sitios donde se incluye. Se podría sustituir perfectamente por una función que comprobara si cada uno de los géneros pertenece, o no, al conjunto.

JJ commented on src/model/conjunto_generos.js:
```diff
@@ -0,0 +1,53 @@
+    * @returns {Set} - Conjunto de géneros global. Los valore del conjunto son
+    * strings
+    */
+   static get GENEROS(){
```
Esto es simplemente una constante. Para empezar, no debes usar literales, también lo hemos repetido muchas veces. Literales devueltos por una función, peor todavía.

JoseCarlosPPK replied:
> Esto es simplemente una constante. Para empezar, no debes usar literales, también lo hemos repetido muchas veces. Literales devueltos por una función, peor todavía.


Antes de continuar, me gustaría debatir esto @JJ y @fjgallardo00  si se quiere animar (si es una clase, una función, constante, etc.). Antes que nada, ¿esto se realiza sobre el propio PR, o se abre un isse, o se comenta sobre el propio issue #10?

JJ reviewed (changes requested):
El problema principal es el código que no se ha desarrollado siguiendo la metodología repetida en prácticamente todas las clases; adicionalmente, se deben tratar de seguir buenas prácticas en el código desarrollado y hacerlo lo más simple posible. No usar una función cuando basta una constante, y no usar una clase cuando basta una función

JoseCarlosPPK commented:
> > Listo para revisión @fjgallardo00 @JJ! He quitado el setter y el código de asginación al atributo solo lo ejecuta el constructor. Además como mejora (pienso), ningún objeto de la clase puede modificar el atributo de clase GENEROS. Para los libros he quitado el atributo _descripcion_ que no se menciona en ningún issue y no es necesario para el PMV.
> > El issue #12 he reformulado el título para que exprese un problema, no sé si el error estaba ahí, en la expresión. El problema es que se han de modelar los libros porque la lógica de negocio los usa para sus cálculos y finalmente como recomendaciones (esto está expresado en el issue)
> 
> Repetidamente os he dicho que reformular un issue es una mala práctica. La buena práctica es cerrarlo cuando se borra el código que se ha creado con él y crear un nuevo issue.

Pues tengo un problema. Si tuviera que crear otro issue sería muy parecido al que ya está. Pensaba que el error estaba en cómo había formulado o expresado el issue (el título). Me gustaría saber el porqué no representa un problema el issue #12

JJ commented:
No he dicho que no represente un problema. He dicho que reescribir un issue es una mala práctica. Hay que cerrarlo, borrar el código y hacer uno nuevo.

JoseCarlosPPK committed:
Delete import innecesario; closes #12

JoseCarlosPPK committed:
Delete import innecesario;avanzo #15

JoseCarlosPPK committed:
Delete setter de Usuario; closes #15

Los setter en el modelo del problema no son necesarios. No se añade funcionalidad

JoseCarlosPPK commented:
Otra pregunta @JJ es si tengo que tener un issue general, que represente al objetivo por así decirlo para añadir el fichero iv.yaml . O tal vez para este no haga falta

JJ commented:
TL;DR: *siempre* que se aporte código al repositorio, hay que hacerlo siguiendo un issue.

> Otra pregunta @JJ es si tengo que tener un issue general, que represente al objetivo por así decirlo para añadir el fichero iv.yaml . O tal vez para este no haga falta

Esa no es la pregunta. La pregunta es si el issue plantea un problema. Los issues *siempre* tienen que plantear un problema.

También la pregunta es "si pongo esto así te quedarás satisfecho". No, no me quedaré satisfecho. Los issues se crean no porque a mi me guste o me haya dado por ahí, y no se plantean como un problema para echar para atrás todo lo que no me parezca un problema. Se plantean porque la ingeniería es resolver problemas (dicho el primer día de clase) y la única forma de saber si realmente se ha resuelto un problema es plantear un issue, escribir código para resolverlo, y escribir un test que compruebe que efectivamente el código resuelve el problema.

Por otro lado, la pregunta es "si hago esto así, si estará bien o no". Y esto también os lo respondí el primer día. La asignatura, y la ingeniería en general, te tiene que decir dos cosas: qué tengo que hacer ahora, y si lo que hago está bien. Para lo primero, está la progresión de los objetivos; para lo segundo, la asignatura tiene las revisiones de código. Así que hasta que no escribas issue, escribas el código correspondiente, y hagas el test (o en este objetivo aprobación por @fjgallardo00 y por mi parte) o se revise, no sabré si está bien o no.

Entiendo que queráis optimizar vuestro tiempo y no hacer cosas que están mal. Pero debéis de entender que para no hacer cosas que están mal tenéis que haber seguido las directrices y metodologías que están en clase, y la metodología de dar clase implica que si no son correctas o no contribuyen al objetivo, se revisan y se solicita que se cambien. Para optimizar *mi* tiempo, desde luego, lo más adecuado sería que intentárais hacer las cosas tras mirar los errores frecuentes y las explicaciones que se han hecho en clase, en este caso sobre el concepto de issue. Pero es legítimo que hagáis lo que veáis y yo tenga que revisarlo; aparte, es mi obligación hacerlo y es como aporto valor a vuestro trabajo.

JoseCarlosPPK committed:
Reset objetivo 2; borro todo lo anterior y vuelta a empezar

closes #10 closes #11 closes #12 closes #13 closes #15

JoseCarlosPPK committed:
Add fichero iv.yaml y elegido lenguaje de programación

Avanzo #16 y closes #8
